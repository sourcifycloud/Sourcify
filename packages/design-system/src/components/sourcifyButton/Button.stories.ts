import sourcifyButton from './Button.vue';
import { action } from '@storybook/addon-actions';
import type { StoryFn } from '@storybook/vue';

export default {
	title: 'Atoms/Button',
	component: sourcifyButton,
	argTypes: {
		type: {
			control: 'select',
			options: ['primary', 'secondary', 'tertiary', 'success', 'warning', 'danger'],
		},
		size: {
			control: {
				type: 'select',
				options: ['mini', 'small', 'medium', 'large', 'xlarge'],
			},
		},
		float: {
			type: 'select',
			options: ['left', 'right'],
		},
	},
	parameters: {
		design: {
			type: 'figma',
			url: 'https://www.figma.com/file/DxLbnIyMK8X0uLkUguFV4n/sourcify-design-system_v1?node-id=5%3A1147',
		},
	},
};

const methods = {
	onClick: action('click'),
};

const Template: StoryFn = (args, { argTypes }) => ({
	props: Object.keys(argTypes),
	components: {
		sourcifyButton,
	},
	template: '<sourcify-button v-bind="$props" @click="onClick" />',
	methods,
});

export const Button = Template.bind({});
Button.args = {
	label: 'Button',
};

const AllSizesTemplate: StoryFn = (args, { argTypes }) => ({
	props: Object.keys(argTypes),
	components: {
		sourcifyButton,
	},
	template: `<div>
		<sourcify-button v-bind="$props" size="large" @click="onClick" />
		<sourcify-button v-bind="$props" size="medium" @click="onClick" />
		<sourcify-button v-bind="$props" size="small" @click="onClick" />
		<sourcify-button v-bind="$props" :loading="true" @click="onClick" />
		<sourcify-button v-bind="$props" :disabled="true" @click="onClick" />
	</div>`,
	methods,
});

const AllColorsAndSizesTemplate: StoryFn = (args, { argTypes }) => ({
	props: Object.keys(argTypes),
	components: {
		sourcifyButton,
	},
	template: `<div>
		<sourcify-button v-bind="$props" size="large" type="primary" @click="onClick" />
		<sourcify-button v-bind="$props" size="large" type="secondary" @click="onClick" />
		<sourcify-button v-bind="$props" size="large" type="tertiary" @click="onClick" />
		<sourcify-button v-bind="$props" size="large" type="success" @click="onClick" />
		<sourcify-button v-bind="$props" size="large" type="warning" @click="onClick" />
		<sourcify-button v-bind="$props" size="large" type="danger" @click="onClick" />
		<br/>
		<br/>
		<sourcify-button v-bind="$props" size="medium" type="primary" @click="onClick" />
		<sourcify-button v-bind="$props" size="medium" type="secondary" @click="onClick" />
		<sourcify-button v-bind="$props" size="medium" type="tertiary" @click="onClick" />
		<sourcify-button v-bind="$props" size="medium" type="success" @click="onClick" />
		<sourcify-button v-bind="$props" size="medium" type="warning" @click="onClick" />
		<sourcify-button v-bind="$props" size="medium" type="danger" @click="onClick" />
		<br/>
		<br/>
		<sourcify-button v-bind="$props" size="small" type="primary" @click="onClick" />
		<sourcify-button v-bind="$props" size="small" type="secondary" @click="onClick" />
		<sourcify-button v-bind="$props" size="small" type="tertiary" @click="onClick" />
		<sourcify-button v-bind="$props" size="small" type="success" @click="onClick" />
		<sourcify-button v-bind="$props" size="small" type="warning" @click="onClick" />
		<sourcify-button v-bind="$props" size="small" type="danger" @click="onClick" />
	</div>`,
	methods,
});

export const Primary = AllSizesTemplate.bind({});
Primary.args = {
	type: 'primary',
	label: 'Button',
};

export const Secondary = AllSizesTemplate.bind({});
Secondary.args = {
	type: 'secondary',
	label: 'Button',
};

export const Tertiary = AllSizesTemplate.bind({});
Tertiary.args = {
	type: 'tertiary',
	label: 'Button',
};

export const Success = AllSizesTemplate.bind({});
Success.args = {
	type: 'success',
	label: 'Button',
};

export const Warning = AllSizesTemplate.bind({});
Warning.args = {
	type: 'warning',
	label: 'Button',
};

export const Danger = AllSizesTemplate.bind({});
Danger.args = {
	type: 'danger',
	label: 'Button',
};

export const Outline = AllColorsAndSizesTemplate.bind({});
Outline.args = {
	outline: true,
	label: 'Button',
};

export const Text = AllColorsAndSizesTemplate.bind({});
Text.args = {
	text: true,
	label: 'Button',
};

export const WithIcon = AllSizesTemplate.bind({});
WithIcon.args = {
	label: 'Button',
	icon: 'plus-circle',
};

export const Square = AllColorsAndSizesTemplate.bind({});
Square.args = {
	label: '48',
	square: true,
};
