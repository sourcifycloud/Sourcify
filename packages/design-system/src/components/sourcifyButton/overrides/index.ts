export { default as sourcifyElButton } from './ElButton.vue';
