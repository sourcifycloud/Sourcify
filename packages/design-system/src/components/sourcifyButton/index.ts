import sourcifyButton from './Button.vue';

export default sourcifyButton;
