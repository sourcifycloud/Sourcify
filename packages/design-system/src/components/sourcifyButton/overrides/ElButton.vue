<template>
	<sourcify-button ref="button" v-bind="attrs" v-on="$listeners">
		<slot />
	</sourcify-button>
</template>

<script lang="ts">
import { defineComponent } from 'vue';
import sourcifyButton from '../Button.vue';

const classToTypeMap = {
	'btn--cancel': 'secondary',
	'el-picker-panel__link-btn': 'secondary',
};

export default defineComponent({
	components: {
		sourcifyButton,
	},
	computed: {
		attrs() {
			let type = this.$attrs.type || 'primary';

			/* Element UI Button can have 'text' or 'info' type which is not supported by sourcify-button
			 so render it as 'secondary' or 'tertiary' */
			if (type === 'text') {
				type = 'secondary';
			}

			if (type === 'info') {
				type = 'tertiary';
			}

			Object.entries(classToTypeMap).forEach(([className, mappedType]) => {
				if (this.$refs.button && (this.$refs.button as Vue).$el.classList.contains(className)) {
					type = mappedType;
				}
			});

			delete this.$attrs.type;

			return {
				...this.$attrs,
				type,
			};
		},
	},
});
</script>
