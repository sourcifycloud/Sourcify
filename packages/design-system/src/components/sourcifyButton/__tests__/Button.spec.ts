import { render } from '@testing-library/vue';
import sourcifyButton from '../Button.vue';
import ElButton from '../overrides/ElButton.vue';

const slots = {
	default: 'Button',
};
const stubs = ['sourcify-spinner', 'sourcify-icon'];

describe('components', () => {
	describe('sourcifyButton', () => {
		it('should render correctly', () => {
			const wrapper = render(sourcifyButton, {
				slots,
				stubs,
			});
			expect(wrapper.html()).toMatchSnapshot();
		});

		describe('props', () => {
			describe('loading', () => {
				it('should render loading spinner', () => {
					const wrapper = render(sourcifyButton, {
						props: {
							loading: true,
						},
						slots,
						stubs,
					});
					expect(wrapper.html()).toMatchSnapshot();
				});
			});

			describe('icon', () => {
				it('should render icon button', () => {
					const wrapper = render(sourcifyButton, {
						props: {
							icon: 'plus-circle',
						},
						slots,
						stubs,
					});
					expect(wrapper.html()).toMatchSnapshot();
				});
			});

			describe('square', () => {
				it('should render square button', () => {
					const wrapper = render(sourcifyButton, {
						props: {
							square: true,
							label: '48',
						},
						stubs,
					});
					expect(wrapper.html()).toMatchSnapshot();
				});
			});
		});

		describe('overrides', () => {
			it('should use default (`primary`) type when no type is given', () => {
				const wrapper = render(ElButton, {
					props: {
						icon: 'plus-circle',
					},
					slots,
					stubs,
				});

				expect(wrapper.html()).toMatchSnapshot();
			});

			it('should use given (`secondary`) type', () => {
				const wrapper = render(ElButton, {
					props: {
						icon: 'plus-circle',
						type: 'secondary',
					},
					slots,
					stubs,
				});

				expect(wrapper.html()).toMatchSnapshot();
			});

			it('should render as `secondary` when `text` is given as type', () => {
				const wrapper = render(ElButton, {
					props: {
						icon: 'plus-circle',
						type: 'text',
					},
					slots,
					stubs,
				});

				expect(wrapper.html()).toMatchSnapshot();
			});

			it('should render as `tertiary` when `info` is given as type', () => {
				const wrapper = render(ElButton, {
					props: {
						icon: 'plus-circle',
						type: 'info',
					},
					slots,
					stubs,
				});

				expect(wrapper.html()).toMatchSnapshot();
			});
		});
	});
});
