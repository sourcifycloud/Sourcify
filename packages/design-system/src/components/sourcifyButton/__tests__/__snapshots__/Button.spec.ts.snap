// Vitest Snapshot v1

exports[`components > sourcifyButton > overrides > should render as \`secondary\` when \`text\` is given as type 1`] = `"<button aria-live=\\"polite\\" class=\\"button button secondary medium icon\\" icon=\\"plus-circle\\"><span class=\\"icon\\"><sourcify-icon-stub icon=\\"plus-circle\\" size=\\"medium\\"></sourcify-icon-stub></span><span>Button</span></button>"`;

exports[`components > sourcifyButton > overrides > should render as \`tertiary\` when \`info\` is given as type 1`] = `"<button aria-live=\\"polite\\" class=\\"button button tertiary medium icon\\" icon=\\"plus-circle\\"><span class=\\"icon\\"><sourcify-icon-stub icon=\\"plus-circle\\" size=\\"medium\\"></sourcify-icon-stub></span><span>Button</span></button>"`;

exports[`components > sourcifyButton > overrides > should use default (\`primary\`) type when no type is given 1`] = `"<button aria-live=\\"polite\\" class=\\"button button primary medium icon\\" icon=\\"plus-circle\\"><span class=\\"icon\\"><sourcify-icon-stub icon=\\"plus-circle\\" size=\\"medium\\"></sourcify-icon-stub></span><span>Button</span></button>"`;

exports[`components > sourcifyButton > overrides > should use given (\`secondary\`) type 1`] = `"<button aria-live=\\"polite\\" class=\\"button button secondary medium icon\\" icon=\\"plus-circle\\"><span class=\\"icon\\"><sourcify-icon-stub icon=\\"plus-circle\\" size=\\"medium\\"></sourcify-icon-stub></span><span>Button</span></button>"`;

exports[`components > sourcifyButton > props > icon > should render icon button 1`] = `"<button aria-live=\\"polite\\" class=\\"button button primary medium icon\\"><span class=\\"icon\\"><sourcify-icon-stub icon=\\"plus-circle\\" size=\\"medium\\"></sourcify-icon-stub></span><span>Button</span></button>"`;

exports[`components > sourcifyButton > props > loading > should render loading spinner 1`] = `"<button disabled=\\"disabled\\" aria-busy=\\"true\\" aria-live=\\"polite\\" class=\\"button button primary medium loading icon\\"><span class=\\"icon\\"><sourcify-spinner-stub size=\\"medium\\" type=\\"dots\\"></sourcify-spinner-stub></span><span>Button</span></button>"`;

exports[`components > sourcifyButton > props > square > should render square button 1`] = `
"<button aria-live=\\"polite\\" class=\\"button button primary medium square\\">
  <!----><span>48</span></button>"
`;

exports[`components > sourcifyButton > should render correctly 1`] = `
"<button aria-live=\\"polite\\" class=\\"button button primary medium\\">
  <!----><span>Button</span></button>"
`;
